File: toaster.py
from time import sleep

class Toaster:

    def __init__(self, farbe, schaechte):

        self.farbe = farbe
        self.schaechte = schaechte

        self.toastzeit = 0
        self.anzahl_toasts = 0
        self.toasts_zustand = 0 #0 Ungetoastet, 1 leicht getoasted, 2 stark getoasted, 3 verbrannt

    def toast_reintun(self, anzahl):
        if anzahl <= self.schaechte:
            self.anzahl_toasts = anzahl
            self.toasts_zustand = 0
        else:
            print("Nicht genug Platz!")

    def toasten(self):
        if self.anzahl_toasts == 0:
            print("Geht nicht. Kein Toast drin.")
        else:
            sleep(self.toastzeit)
            self.toasts_zustand = 2
            print("Fertig")

    def toast_auswerfen(self):
        print(f"{self.anzahl_toasts} Toast ausgeworfen. Toaster ist leer.")
        self.anzahl_toasts = 0

    def zeit_einstellen(self, zeit):
        self.toastzeit = zeit

mein_toaster = Toaster("Rot", 2)
mein_toaster.zeit_einstellen(5)

print(mein_toaster.farbe)

mein_toaster.toast_reintun(2)
mein_toaster.toasten()
mein_toaster.toast_auswerfen()

File: TypeScript/modules/toaster.ts
import { toastZustandAlsString, delay } from "./util.ts"

export class Toaster {

    // Deklarieren der props. Standardmäßig als private außer jene die von Subklassen benötigt werden.
    // Die werden als protected angegeben um Subklasen wie SuperToaster den Zugriff zu gewähren.
    // Readonly weil diee Props nicht veränderbar sein sollen nach der Initialisierung
    private readonly farbe: string
    private readonly schaechte: number
    protected toastZeit = 0
    protected anzahlToast = 0
    protected toastZustand = 0

    // Constructor welcher Farbe und Schächte direkt bei der Erstellung setzt (unveränderbar im Nachhinein)
    constructor(farbe: string, schaechte: number) {
        this.farbe = farbe;
        this.schaechte = schaechte;
        
    }

    // Einfache Methoden
    toastReintun(anzahl: number): boolean {
        if (anzahl <= this.schaechte) {
            this.anzahlToast = anzahl;
            console.log(`${this.anzahlToast} Toast reingetan.`);
            return true
        } else {
            console.log("Nicht genug Platz!");
            return false
        }
    }

    async toasten(): Promise<void> {
        if (this.anzahlToast == 0) {
            console.log("Geht nicht. Kein Toast drin.");
        } else {
            console.log("Toasten...");
            // Besonderheit: JavaScript hat kein time.sleep() wie Python weshalb auf setInterval zugegriffen werden muss.
            // Allerdings ist setInterval nicht Async weshalb andere Teile des Programmes während der Verzögerung weiterlaufen würden.
            // Darum wurde hier setInterval ausgelagert und mit Hilfe von Async Await asynchron gemacht.
            await delay(this.toastZeit * 1000)
            if (this.toastZeit >= 0) {
                this.toastZustand = 0
            } else if (this.toastZeit <= 15) {
                this.toastZustand = 1
            } else if (this.toastZeit > 15) {
                this.toastZustand = 2
            } else if (this.toastZeit > 30) {
                this.toastZustand = 3
            }
            console.log(`Fertig. Toast ist: ${toastZustandAlsString(this.toastZustand)}`);
        }
    }

    toastAuswerfen(): void {
        console.log(`${this.anzahlToast} Toast ausgeworfen. Toaster ist nun leer.\n`);
        this.anzahlToast = 0;
        this.toastZustand = 0;
    }

    zeitEinstellen(zeit: number): void {
        this.toastZeit = zeit;
        console.log(`Zeit eingestellt auf ${zeit} Sekunden`);
    }

    getFarbe(): string {
        return this.farbe;
    }

    getSchaechte(): number {
        return this.schaechte;
    }

    getZeit(): number {
        return this.toastZeit;
    }

}

